etc/ha_healthchecker/ha_healthchecker.py

```diff
@@ -0,0 +1,748 @@
+
+    def _is_alarmed_timeout(self, obj):
+        over_time = self._parse_isotime(
+            obj.alarmed_at) + datetime.timedelta(
+            days=2)
+        current_time = datetime.datetime.utcnow().replace(tzinfo=iso8601.UTC)
+        return current_time > over_time
+
+    def _get_service_status(self, service):
+        # gearman is controlled by zuul itself. It's not under systemd.
+        if service == 'gearman':
+            return 'up' if os.path.exists(
+                '/run/gearman/server.pid') else 'down'
+
+        # timer tasks are handled by crontab
+        if service in ['zuul-timer-tasks', 'nodepool-timer-tasks']:
+            service = 'cron'
+
+        cmd = "systemctl status {srvc}".format(srvc=service)
+        try:
+            # 0 means OK
+            # if using status command, 0 means active, non-zero means error
+            # status.
+            subprocess.check_output(cmd.split(), stderr=subprocess.STDOUT)
+            LOG.debug("Run CMD: %(cmd)s", {'cmd': cmd})
+            return 'up'
+        except subprocess.CalledProcessError as e:
+            LOG.error("Failed to check %(srvc)s service: "
+                      "%(err)s %(out)s", {'srvc': service, 'err': e,
+                                          'out': e.output})
+            return 'down'
+
+
+class Refresher(Base):
+    def __init__(self, zk, heartbeat_timeout):
+        super(Refresher, self).__init__(zk, heartbeat_timeout)
+
+    def _local_node_service_process(self, node_obj):
+        service_objs = self.zk.list_services(node_name_filter=node_obj.name)
+        for service_obj in service_objs:
+            self._refresh_service(service_obj, node_obj)
+
+        self._report_heart_beat(node_obj)
+        if node_obj.alarmed and node_obj.role == 'slave':
+            self.zk.update_node(node_obj.name, alarmed=False)
+
+    def _refresh_service(self, service_obj, node_obj):
+        cur_status = self._get_service_status(service_obj.name)
+        LOG.debug("Service %(name)s check result: %(status)s",
+                  {'name': service_obj.name, 'status': cur_status})
+        update_dict = {}
+        if cur_status == 'up':
+            if service_obj.status != 'up':
+                update_dict['status'] = 'up'
+                update_dict['restarted'] = False
+                update_dict['alarmed'] = False
+        else:
+            if not service_obj.restarted:
+                update_dict['status'] = 'restarting'
+                update_dict['restarted'] = True
+            else:
+                update_dict['status'] = 'down'
+        if update_dict:
+            self.zk.update_service(service_obj.name, node_obj.name,
+                                   **update_dict)
+
+        if cur_status != 'up':
+            LOG.debug(
+                "Service %(name)s is abnormal. It's status is %(status)s.",
+                {'name': service_obj.name,
+                 'status': cur_status.upper()})
```

Comment:
if the service is abnormal, it's always `down`, this error message should be rewrote.

Reply:
Correct, each loop will report this message till this node to be set to 'down', we just want a single message? If yes, we can introduce the restarted or alarmed here, or just log when svr.status == 'restarting'